### PR TITLE
[llm] Add convert_model api

### PR DESCRIPTION
**Pull request by ACupofAir**
## Description
combine #8235 and #8228  to one api `convert_model`

### 1. User API changes

Basic usage of convert a model to  int4-quantized model

```python
from bigdl.llm.ggml import convert_model
model_path = .'/models/vicuna-13b-1.1'
convert_model(model_path=model_path, output_path=None, model_family='llama')
```

### 2. Summary of the change 

* add `convert_model` api for bigdl.llm.ggml
* fix a path name replace problem of pathlib.Path in `quantize.py`

### 3. How to test?
successfully test on vicuna-13b-1.1

### CODE REVIEW

**Comment by jason-dai** [comment]
> ```python
> from bigdl.llm.ggml import convert_model
> model_path = .'/models/vicuna-13b-1.1'
> convert_model(model_path=model_path, output_path=None, model_family='llama')
> ```

- Change `model_path` to `input_path`
- Do we want to have a default value for `model_family`? Since bigdl-llm is supposed to support general llm, a default value may not make sense @TheaperDeng

**Comment by jason-dai** [comment]
And it's not clear to me what the expected input model format is; do we check if the input format is valid?

**Comment by TheaperDeng** [comment]
We need to specify what 

> > ```python
> > from bigdl.llm.ggml import convert_model
> > model_path = .'/models/vicuna-13b-1.1'
> > convert_model(model_path=model_path, output_path=None, model_family='llama')
> > ```
> 
> * Change `model_path` to `input_path`
> * Do we want to have a default value for `model_family`? Since bigdl-llm is supposed to support general llm, a default value may not make sense
> * How does the user specify the output type?
> 
> In addition, we need to unify the `model_family` handling in bigdl-llm, including:
> 
> * Exact name of the families: e.g., people are using "gptneox" or "redpajama" in different places; maybe we should define these families in a central places
> * Is there a default model family - probably not
> * Check if the input model family is valid.
> 
> @TheaperDeng

- We will use `gptneox` instead of `redpajama`, currently the later one is only used as some branch-name or repo-name we build at very early stage, which will not be exposed to users. After https://github.com/intel-analytics/BigDL/pull/8246, I don't think we will see any `redpajama` in bigdl repo.
- We may leave the model family a required parameter for now. In next PR, we may directly infer the family from the model dir's config.json (if existed)
- We need to add detailed structure requirement in our comment **in this PR** (including the file structure for tokenizer, modelbin, and other files)

**Comment by TheaperDeng** [comment]
what we support now is what users directly pull from huggingface hub.

Like these
https://huggingface.co/bigscience/bloomz-7b1/tree/main
https://huggingface.co/decapoda-research/llama-7b-hf/tree/main

**Comment by jason-dai** [comment]
> what we support now is what users directly pull from huggingface hub.
> 
> Like these https://huggingface.co/bigscience/bloomz-7b1/tree/main https://huggingface.co/decapoda-research/llama-7b-hf/tree/main

Need to document how the input can be obtained (e.g., in the python doc, a model convert cli tool, etc.)

**Comment by jason-dai** [comment]
> * We will use `gptneox` instead of `redpajama`, currently the later one is only used as some branch-name or repo-name we build at very early stage, which will not be exposed to users. After [[LLM]Support for linux package (llama, NeoX) & quantize (llama) #8246](https://github.com/intel-analytics/BigDL/pull/8246), I don't think we will see any `redpajama` in bigdl repo.

I think the command line tool currently uses `redpajama` @hoshibara 

We need to have a central utility to manage all the model families

File: python/llm/src/bigdl/llm/ggml/convert_model.py [hunk 1]
```diff
@@ -4,14 +4,14 @@
...
-def convert_model(model_path: str,
+def convert_model(input_path: str,
                   output_path: str = None,
                   model_family: str = 'llama',
                   dtype: str = 'q4_0'):
     """
     Convert Hugging Face llama-like / gpt-neox-like / bloom-like model to lower precision
 
-    :param model_path: str, path of model, for example `./llama-7b-hf`.
+    :param input_path: str, path of model, for example `./llama-7b-hf`.
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/ggml/convert_model.py:14]: Need to document what is expected in the `input_path` folder, and how to obtain the folder for the user

Comment (ACupofAir) [reply python/llm/src/bigdl/llm/ggml/convert_model.py:14]: - I will add the document of input_path format in next pr, this pr  is being relied upon by other developers and should be merged as soon as possible. 
- I add an issue #8254  to track the document format problem.

File: python/llm/src/bigdl/llm/ggml/convert_model.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+from bigdl.llm.ggml.convert import _convert_to_ggml
+from bigdl.llm.ggml.quantize import quantize
+from pathlib import Path
+import time
+
+
+def convert_model(input_path: str,
+                  output_path: str = None,
+                  model_family: str = 'llama',
+                  dtype: str = 'q4_0'):
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/ggml/convert_model.py:10]: `dtype` should be `int4` by default, which maps to `int4_0`;

File: python/llm/src/bigdl/llm/ggml/convert_model.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            If you don't specify this parameter, quantized model will be saved in
+            the same directory as the input and just replace precision with quantize_type
+            like `./ggml-model-q4_0.bin`.
+    :param model_family: Which model family your input model belongs to. Default to `llama`.
+            Now only `llama`/`bloomz`/`gptneox` are supported.
+    :param dtype: Which quantized precision will be converted
+            Now only int4 supported.
+    """
+
+    if dtype == 'int4':
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/ggml/convert_model.py:25]: first change `dtype` to lower case first

File: python/llm/src/bigdl/llm/ggml/convert_model.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+from bigdl.llm.ggml.convert import _convert_to_ggml
+from bigdl.llm.ggml.quantize import quantize
+from pathlib import Path
+import time
+
+
+def convert_model(input_path: str,
+                  output_path: str = None,
+                  model_family: str = 'llama',
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/ggml/convert_model.py:9]: remove default value 'llama'

File: python/llm/src/bigdl/llm/ggml/convert_model.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    :param output_path: Save path of output quantized model. Default to `None`.
+            If you don't specify this parameter, quantized model will be saved in
+            the same directory as the input and just replace precision with quantize_type
+            like `./ggml-model-q4_0.bin`.
+    :param model_family: Which model family your input model belongs to. Default to `llama`.
+            Now only `llama`/`bloomz`/`gptneox` are supported.
+    :param dtype: Which quantized precision will be converted
+            Now only int4 supported.
+    """
+
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/ggml/convert_model.py:24]: you need to check if the input values are valid

File: python/llm/src/bigdl/llm/ggml/convert_model.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+                  model_family: str = 'llama',
+                  dtype: str = 'int4'):
+    """
+    Convert Hugging Face llama-like / gpt-neox-like / bloom-like model to lower precision
+
+    :param input_path: str, path of model, for example `./llama-7b-hf`.
+    :param output_path: Save path of output quantized model. Default to `None`.
+            If you don't specify this parameter, quantized model will be saved in
+            the same directory as the input and just replace precision with quantize_type
+            like `./ggml-model-q4_0.bin`.
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/ggml/convert_model.py:18]: I don't think we should support `None`; we may support model meta info in future @TheaperDeng ?

File: python/llm/src/bigdl/llm/ggml/convert_model.py (hunk 5 above)
Comment (ACupofAir) [reply python/llm/src/bigdl/llm/ggml/convert_model.py:24]: * The `input_path` has checked in `_convert_to_ggml` function
* The `dtype` has checked in `quantize` function

File: python/llm/src/bigdl/llm/ggml/convert_model.py (hunk 1 above)
Comment (TheaperDeng) [reply python/llm/src/bigdl/llm/ggml/convert_model.py:14]: I will take this issue for input path format description

File: python/llm/src/bigdl/llm/ggml/convert_model.py (hunk 6 above)
Comment (TheaperDeng) [reply python/llm/src/bigdl/llm/ggml/convert_model.py:18]: yeah a good way for future is to require a dir path here.

File: python/llm/src/bigdl/llm/ggml/convert_model.py (hunk 5 above)
Comment (jason-dai) [reply python/llm/src/bigdl/llm/ggml/convert_model.py:24]: The guideline is always to check and report error to the user as early as possible

**Review by TheaperDeng — APPROVED** [review]
TODOs are recorded in https://github.com/intel-analytics/BigDL/issues/8254

**Comment by TheaperDeng** [comment]
Some further improvement will be added in https://github.com/intel-analytics/BigDL/issues/8254. Merge this PR for now to speed-up some dependent work.

File: python/llm/src/bigdl/llm/ggml/quantize.py [hunk 7]
```diff
@@ -38,7 +39,7 @@
                           "q8_0": 7}
 
 _quantize_type = {"llama": _llama_quantize_type,
-                  "bloomz": _bloomz_quantize_type,
+                  "bloom": _bloom_quantize_type,
                   "gptneox": _gptneox_quantize_type}
 
 _valid_types = set(list(_llama_quantize_type.keys()) + list(_bloomz_quantize_type.keys()) +
```
Comment (Oscilloscope98) [inline python/llm/src/bigdl/llm/ggml/quantize.py:45]: `_bloomz_quantize_type.keys()` => `_bloom_quantize_type.keys()`
Seems that you forgot to change here

**Pull request merged**